### PR TITLE
fix(send): chunker skips empty chunks so blank-line boot prompts deliver

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- Coalesce whitespace-only chunks into adjacent chunks so `cmux set-buffer` is never called with empty/blank text during chunked prompt delivery.
- Apply the fix to both `src/server.ts` and the mirrored app-server runtime chunker.
- Add a regression test for a multi-paragraph inline `spawn_agent` boot prompt with a blank line at the chunk boundary, asserting no whitespace-only chunks and exact prompt preservation.

## Test plan
- `graphify query "chunkTerminalInput deliverInputChunks SEND_INPUT_CHUNK_THRESHOLD empty chunk blank line set-buffer" --budget 3000`
- `bun test tests/server-agent-tools.test.ts -t "spawn_agent delivers inline prompts with blank lines without empty chunks"`
- `bun test tests/server-agent-tools.test.ts -t "spawn_agent sends inline prompt after the agent is ready"`
- `bun test tests/server.test.ts -t "send_input chunks long text transparently before sending"`
- `bun run test`
- `bun run typecheck`
- `bun run build`
- `coderabbit review --agent` (second pass: 0 findings)

---

> [!NOTE]
> **Low Risk**
> Localized change to chunk assembly plus a test; behavior for prompts without blank-line-only chunks should match the prior newline-first splitting.
> 
> **Overview**
> Fixes **chunked terminal/prompt delivery** when a newline-aligned split leaves a **whitespace-only segment** (e.g. a blank line between paragraphs). Those segments used to become their own chunks and could fail **`set-buffer`** / paste delivery during **`spawn_agent`** boot prompts.
> 
> **`chunkTerminalInput`** now builds raw line-bounded chunks first, then **carries whitespace-only pieces forward**—merging them with the next non-empty chunk (re-splitting if needed) or **appending to the last chunk**—so every emitted chunk has non-empty trimmed content while **joining chunks still reproduces the original text**.
> 
> The same logic is applied in **`src/server.ts`** and the mirrored **`src/app-server-runtime.ts`**. A regression test covers a long inline **`spawn_agent`** prompt with `\n\n` at a chunk boundary, asserting multiple chunks, no empty chunks, size limits, and exact prompt preservation.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7452e9cb518c5c3f51b8cf6c694fa371e1ebcecb. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fix `chunkTerminalInput` to skip whitespace-only chunks so blank-line boot prompts deliver
> - Rewrites the chunking algorithm in [app-server-runtime.ts](https://github.com/EtanHey/cmuxlayer/pull/196/files#diff-b06e09ff3b0e74eac8e2e15c6e8907fcab61898d9cf49c2f6a8f6096c32126c3) and [server.ts](https://github.com/EtanHey/cmuxlayer/pull/196/files#diff-8a8ae07582c9d433ec8c2e5c4310ff8901e604f4965c5b90a49117ad46c47595): first slices input at newline boundaries, then merges whitespace-only segments into adjacent text chunks via a `whitespaceCarry` accumulator.
> - Any trailing `whitespaceCarry` is appended to the last non-empty chunk rather than emitted as a standalone empty chunk.
> - Adds a test in [server-agent-tools.test.ts](https://github.com/EtanHey/cmuxlayer/pull/196/files#diff-b02b20318090fbe2fe9736638b282662f0d778996959f4a29828dc0dab2debff) that verifies `spawn_agent` with a blank-line-containing prompt returns `boot_prompt_delivered=true`, emits no empty chunks, and round-trips the full prompt.
> - Behavioral Change: `chunkTerminalInput` now returns `[]` for empty input instead of `['']`, and chunk split boundaries may shift to one character past the first non-whitespace index.
>
> 
>
> <details>
> <summary>📊 <a href="https://app.macroscope.com">Macroscope</a> summarized 7452e9c. 2 files reviewed, 0 issues evaluated, 0 issues filtered, 0 comments posted</summary>
>
> ### 🗂️ Filtered Issues
> No issues evaluated.
>
>
> </details>
>

### CODE REVIEW

**Comment by EtanHey** [comment]
@coderabbitai review

**Comment by EtanHey** [comment]
@codex review

**Comment by EtanHey** [comment]
@cursor @bugbot review

File: src/app-server-runtime.ts [hunk 1]
```diff
@@ -26,22 +26,51 @@ const LAUNCH_SHELL_READY_POLL_MS = 100;
...
+    if (candidate.trim().length === 0) {
+      whitespaceCarry = candidate;
+    } else {
+      chunks.push(candidate);
+    }
+  }
+
+  if (whitespaceCarry && chunks.length > 0) {
+    chunks[chunks.length - 1] += whitespaceCarry;
   }
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/app-server-runtime.ts:75]: Reviewed. This suggestion would reintroduce whitespace-only chunks, which conflicts with the explicit contract for this fix: `chunkTerminalInput` must never emit empty or whitespace-only chunks because those can reach `cmux set-buffer`. For prompts with real content, the regression test verifies `chunks.join("") === prompt` and preserves the embedded blank line. A payload made entirely of whitespace has no adjacent non-whitespace chunk to coalesce into; preserving that would need a separate non-`set-buffer` delivery mode, not this chunker fix.

File: src/server.ts [hunk 2]
```diff
@@ -514,22 +514,51 @@ function applySurfaceWorkingDirectory(
...
+    }
+    if (candidate.trim().length === 0) {
+      whitespaceCarry = candidate;
+    } else {
+      chunks.push(candidate);
+    }
+  }
+
+  if (whitespaceCarry && chunks.length > 0) {
+    chunks[chunks.length - 1] += whitespaceCarry;
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/server.ts:561]: Reviewed. This is a Medium edge case and the proposed behavior conflicts with the bug contract for this PR: the chunker must not emit whitespace-only chunks because `cmux set-buffer` rejects them. The fixed path preserves embedded blank lines by coalescing them into neighboring content chunks and the regression asserts exact round-trip text. Whitespace-only payload preservation would require a separate delivery path that does not call `set-buffer`, so I am not changing the chunker to emit that case here.

**Comment by EtanHey** [comment]
## REVIEW: SIGN-OFF

**Lens:** ponytail (concise / design / readability / correctness)

### Key checks

| # | Invariant | Verdict |
|---|-----------|---------|
| 1 | Chunker never emits empty/whitespace-only chunks → `set-buffer` never gets blank text | **PASS** — coalesce loop skips `trim().length === 0` segments; trailing carry appended to last chunk only when non-empty chunks exist |
| 2 | Content preserved — blank line between paragraphs survives in delivered text | **PASS** — `chunks.join("") === prompt` + `toContain("\n\n")` in new test; manual probe on `500a + \n\n + 600b` round-trips |
| 3 | No regression (single-chunk, no-blank-line, long-prompt) | **PASS** — targeted tests green; full suite **960/960** on branch |
| 4 | HELD test (would fail pre-fix) | **PASS** — mock `set-buffer` throws on `trim().length === 0`; pre-fix emits whitespace-only chunk at raw index 1 for the `\n\n` boundary case → stranded spawn; post-fix delivers `boot_prompt_delivered: true` |

### Design notes (non-blocking)

- Two-phase split (newline-bounded raw → whitespace carry merge) is readable; `whitespaceCarry` intent is clear.
- Mirrored fix in `server.ts` + `app-server-runtime.ts` matches existing duplication pattern.
- Re-split when `carry + chunk > chunkSize` correctly handles oversized merges; `<= 501` bound in test covers the `firstTextIndex + 1` edge.

### Nits (informational, not blocking)

- Whitespace-only input (e.g. `"\n\n"` alone) now returns `[]` instead of a whitespace chunk — content silently skipped rather than `set-buffer` error. Acceptable for boot-prompt scope; no caller sends whitespace-only prompts today.

**Verdict:** Approve for merge. Do not merge from this review.
